### PR TITLE
Provide limited agent management menu for user permission

### PR DESCRIPTION
**Pull request by donggyu04**
#304 

Users can view the list of public and User's agents from the agent management menu.
(This menu previously only supported for admin.)

image

They can only view the following information about the agents.

image

### CODE REVIEW

File: ngrinder-controller/src/main/java/org/ngrinder/agent/controller/AgentManagerController.java [hunk 1]
```diff
@@ -65,26 +67,28 @@
...
-	 *               agents.
-	 * @param model  model
-	 * @return agent/list
 	 */
 	@RequestMapping({"", "/", "/list"})
-	public String getAll(@RequestParam(value = "region", required = false) final String region, ModelMap model) {
+	@PreAuthorize("hasAnyRole('A', 'S', 'U')")
+	public String getAll(final User user, @RequestParam(value = "region", required = false) final String region, ModelMap model) {
+		final Collection<? extends GrantedAuthority> authorities = getCurrentAuthorities();
 		List<AgentInfo> agents = agentManagerService.getAllVisible();
```
Comment (junoyoon) [inline ngrinder-controller/src/main/java/org/ngrinder/agent/controller/AgentManagerController.java]: What about taking agents as ```Collection<AgentInfo> agents``` from here.
If so, we can just re-use agents variable in the following codes.

File: ngrinder-controller/src/main/java/org/ngrinder/agent/controller/AgentManagerController.java [hunk 2]
```diff
@@ -102,6 +106,40 @@ public boolean apply(AgentInfo agentInfo) {
...
+				return agentRegion.startsWith(region + "_owned_" + userId) ||  region.equals(agentRegion);
+			}
+		}
+		return false;
+	}
+
+	/**
+	 * Check current user is admin or superuser
+	 */
+	private boolean adminOrSuperUser(Collection<? extends GrantedAuthority> authorities) {
```
Comment (junoyoon) [inline ngrinder-controller/src/main/java/org/ngrinder/agent/controller/AgentManagerController.java]: I think this should be isAdminOrSuperUser

File: ngrinder-controller/src/main/java/org/ngrinder/agent/controller/AgentManagerController.java [hunk 3]
```diff
@@ -102,6 +106,40 @@ public boolean apply(AgentInfo agentInfo) {
...
+				return agentRegion == null || "all".equals(agentRegion) || "NONE".equals(agentRegion) || agentRegion.endsWith("_owned_" + userId);
+			} else {
+				return agentRegion.startsWith(region + "_owned_" + userId) ||  region.equals(agentRegion);
+			}
+		}
+		return false;
+	}
+
+	/**
+	 * Check current user is admin or superuser
```
Comment (junoyoon) [inline ngrinder-controller/src/main/java/org/ngrinder/agent/controller/AgentManagerController.java]: Fix to ```Check if the current user...```

File: ngrinder-controller/src/main/webapp/WEB-INF/ftl/agent/list.ftl [hunk 4]
```diff
@@ -36,38 +38,55 @@
...
+				<th><@spring.message "agent.list.version"/></th>
+				<th><@spring.message "agent.list.region"/></th>
+			<@security.authorize access="hasRole('A')">
+				<th class="no-click"><@spring.message "agent.list.approved"/></th>
+			</@security.authorize>
 		</tr>
 		</thead>
 		<tbody>
-		<@list list_items=agents others="table_list" colspan="8"; agent>
+		<#if isAdmin??><#assign column=8/><#else><#assign column=6/></#if>
```
Comment (junoyoon) [inline ngrinder-controller/src/main/webapp/WEB-INF/ftl/agent/list.ftl]: isAdmin includes SuperUser?

File: ngrinder-core/pom.xml [hunk 5]
```diff
@@ -82,6 +82,11 @@
 			<version>1.4.1</version>
 		</dependency>
 
+		<dependency>
+			<groupId>org.springframework.security</groupId>
+			<artifactId>spring-security-web</artifactId>
+			<version>${spring.security.version}</version>
+		</dependency>
```
Comment (junoyoon) [inline ngrinder-core/pom.xml]: Why do we need this lib here?
core should not depend on spring security.

File: ngrinder-core/src/main/java/org/ngrinder/common/util/SpringSecurityUtils.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+     */
+    public static boolean containsAuthority(Collection<? extends GrantedAuthority> authorities, String requiredAuth) {
+        for (GrantedAuthority grantedAuthority : authorities) {
+            if (grantedAuthority.getAuthority().equals(requiredAuth)) {
+                return true;
+            }
+        }
+        return false;
+    }
+}
```
Comment (junoyoon) [inline ngrinder-core/src/main/java/org/ngrinder/common/util/SpringSecurityUtils.java]: I think this file should located in controller not core

File: ngrinder-core/src/main/java/org/ngrinder/common/util/SpringSecurityUtils.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+import java.util.ArrayList;
+import java.util.Collection;
+
+import static org.springframework.security.core.context.SecurityContextHolder.getContext;
+
+/**
+ * Spring Security Utility functions.
+ *
+ * @author DongGyu Lee
```
Comment (junoyoon) [inline ngrinder-core/src/main/java/org/ngrinder/common/util/SpringSecurityUtils.java]: Remove out author

File: ngrinder-controller/src/main/webapp/WEB-INF/ftl/agent/list.ftl (hunk 4 above)
Comment (donggyu04) [reply ngrinder-controller/src/main/webapp/WEB-INF/ftl/agent/list.ftl]: No. only include administrator

**Comment by junoyoon** [comment]
LGTM

**Pull request merged**